File: scratch/record-scalarization-integration-20260916/final/src/Test/ListOpsFFI.cc
#include "purescript.h"
#include <functional>
#include <memory>

namespace {
    struct Node;
    using List = std::shared_ptr<const Node>;
    struct Node { int head; List tail; };

    List cons(int head, List tail) {
        return std::make_shared<Node>(Node{head, tail});
    }

    List range(int start, int end) {
        List result;
        for (int value = end; value >= start; --value) {
            result = cons(value, result);
        }
        return result;
    }

    List filterEvens(List values) {
        List evens;
        while (values) {
            if (values->head % 2 == 0) evens = cons(values->head, evens);
            values = values->tail;
        }
        return evens;
    }

    int foldl(const std::function<int(int, int)>& combine, int acc, List values) {
        while (values) {
            acc = combine(acc, values->head);
            values = values->tail;
        }
        return acc;
    }
}

FOREIGN_BEGIN(Test_ListOpsFFI)
exports["runListOpsFFI"] = [](const boxed& in) -> boxed {
    const auto values = range(1, unbox<int>(in));
    const auto evens = filterEvens(values);
    const int result = foldl([](int sum, int value) { return sum + value; }, 0, evens);
    return result;
};
FOREIGN_END

File: scratch/record-scalarization-integration-20260916/final/src/Test/ListOpsFFI.java
private sealed interface IntList permits Nil, Cons {}
private enum Nil implements IntList { INSTANCE }
private record Cons(int value, IntList tail) implements IntList {}

private static IntList range(int start, int end) {
    IntList result = Nil.INSTANCE;
    for (int current = end; current >= start; current--) {
        result = new Cons(current, result);
    }
    return result;
}

private static IntList filterEvens(IntList list) {
    IntList result = Nil.INSTANCE;
    while (list instanceof Cons cons) {
        if (cons.value() % 2 == 0) result = new Cons(cons.value(), result);
        list = cons.tail();
    }
    return result;
}

private static int foldl(java.util.function.IntBinaryOperator operation, int acc, IntList list) {
    while (list instanceof Cons cons) {
        acc = operation.applyAsInt(acc, cons.value());
        list = cons.tail();
    }
    return acc;
}

public static final java.util.function.Function<Object, Object> runListOpsFFI = Bench.nativeBenchmark(input ->
    foldl((acc, value) -> acc + value, 0, filterEvens(range(1, ((Number) input).intValue()))));
